### PR TITLE
.Net: Processes State Management Part 2

### PR DESCRIPTION
**Pull request by esttenorio**
### Description

- Moving `ProcessStateMetadata` logic to factory/extension to try to keep clean existing core components
- Adding check of step name when adding new steps/processes to process - names must be unique so when applying saved state it propagates properly. Currently there are issues if multiple steps have the same name
- State versioning initial support:
    - Processes with 1:1 step match previous and new version (step name change and/or step logic change) - minor changes on process steps and not in root process itself

Fixing #9358 

#### Out of scope but will be addressed in next PRs
    
- Processes with N:M steps (process flow changed and potentially also step logic changed)
- New samples showcasing how to make use of N:M mapping

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Review by crickman — APPROVED** [review]
I really like the centralized serialization logic.  I'll be able to integrate this to the map-step change much more easily.

I like the depth with which the Step03 sample was updated, but I wonder if there isn't an opportunity for a more targeted starting point to introduce the serialization patterns.  Demonstrating these patterns is a more _real_ case is wonderful evidence on how these patterns scale-out...but I wonder if a more simple sample that only targets serialization and versioning would be useful.

File: dotnet/src/Experimental/Process.Core/Internal/KernelProcessStateMetadataExtension.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                        {
+                            // version mismatch - check if migration logic in place
+                            if (step is ProcessBuilder subprocessBuilder)
+                            {
+                                var sanitizedStepState = SanitizeProcessStateMetadata(savedStateMetadata, subprocessBuilder.Steps.ToList());
+                                sanitizedStateMetadata.StepsState[step.Name] = sanitizedStepState;
+                            }
+                            else if (false)
+                            {
+                                // TODO: Hook up logic for custom step override
```
Comment (crickman) [inline dotnet/src/Experimental/Process.Core/Internal/KernelProcessStateMetadataExtension.cs:78]: Perhaps track _TODO_ with an issue on the backlog>?

**Review by crickman — APPROVED** [review]

File: dotnet/src/Experimental/Process.Core/Internal/KernelProcessStateMetadataExtension.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                            }
+                            else
+                            {
+                                // no compatible state found, migrating id only
+                                sanitizedStateMetadata.StepsState[step.Name] = new KernelProcessStateMetadata()
```
Comment (alliscode) [inline dotnet/src/Experimental/Process.Core/Internal/KernelProcessStateMetadataExtension.cs:83]: In this case are we loosing state data? If so, should we through an exception?

**Review by alliscode — APPROVED** [review]

**Comment by alliscode** [comment]
> I really like the centralized serialization logic. I'll be able to integrate this to the map-step change much more easily.
> 
> I like the depth with which the Step03 sample was updated, but I wonder if there isn't an opportunity for a more targeted starting point to introduce the serialization patterns. Demonstrating these patterns is a more _real_ case is wonderful evidence on how these patterns scale-out...but I wonder if a more simple sample that only targets serialization and versioning would be useful.

I agree but I wonder if the learn site is the place for this tutorial style walkthrough?

**Comment by crickman** [comment]
> > I really like the centralized serialization logic. I'll be able to integrate this to the map-step change much more easily.
> > I like the depth with which the Step03 sample was updated, but I wonder if there isn't an opportunity for a more targeted starting point to introduce the serialization patterns. Demonstrating these patterns is a more _real_ case is wonderful evidence on how these patterns scale-out...but I wonder if a more simple sample that only targets serialization and versioning would be useful.
> 
> I agree but I wonder if the learn site is the place for this tutorial style walkthrough?

Sure, or both

File: dotnet/src/Experimental/Process.Core/Internal/KernelProcessStateMetadataExtension.cs (hunk 2 above)
Comment (esttenorio) [reply dotnet/src/Experimental/Process.Core/Internal/KernelProcessStateMetadataExtension.cs:83]: so the part 3 of these changes is to support custom override/migration so if there was one set that would take over, else only the matching name and id would migrate, and a default step state would be set

**Review by crickman — APPROVED** [review]

**Pull request merged**